### PR TITLE
Generate access Keys and login using keys

### PR DESCRIPTION
**Pull request by geof90**
This PR allows you to generate access keys for your account, so that you can login to CodePush using the access key without going through authentication with the browser, e.g in a CI environment like Visual Studio Online.

To generate an access key,
```
code-push access-key add --des <description>
```

To login with the access key,
```
code-push login --key <accessKey>
```

Added documentation in the CLI readme for these scenarios.

Also added 1 unit test.

### CODE REVIEW

File: cli/script/command-parser.ts [hunk 1]
```diff
@@ -202,16 +218,16 @@ var argv = yargs.usage(USAGE_PREFIX + " <command>")
...
+    .command("register", "Register a new account with a specific CodePush server", (yargs: yargs.Argv) => {
+        isValidCommandCategory = true;
+        isValidCommand = true;
+        yargs.usage(USAGE_PREFIX + " register [serverUrl]")
+            .demand(/*count*/ 1, /*max*/ 2)  // Require one non-optional and one optional argument.
+            .example("register", "Creates a new user account with " + CODE_PUSH_URL)
+            .check((argv: any, aliases: { [aliases: string]: string }): any => isValidCommand);  // Report unrecognized, non-hyphenated command category.
+
+        addCommonConfiguration(yargs);
+    })
```
Comment (shishirx34) [inline cli/script/command-parser.ts]: was this intentional?

Comment (geof90) [inline cli/script/command-parser.ts]: nope it wasn't. thanks for spotting that

File: cli/script/command-executor.ts [hunk 2]
```diff
@@ -438,9 +449,24 @@ function initiateExternalAuthenticationAsync(serverUrl: string, action: string):
 }
 
 function login(command: cli.ILoginCommand): Promise<void> {
-    initiateExternalAuthenticationAsync(command.serverUrl, "login");
+    // Check if one of the flags were provided.
+    if (command.accessKeyName || command.providerName || command.providerUniqueId) {
+        throwForMissingCredentials(command.accessKeyName, command.providerName, command.providerUniqueId);
+        
+        sdk = new AccountManager(command.serverUrl);
+        var accessToken: string = base64.encode(JSON.stringify({ accessKeyName: command.accessKeyName, providerName: command.providerName.toLowerCase(), providerUniqueId: command.providerUniqueId }));
```
Comment (shishirx34) [inline cli/script/command-executor.ts]: not super important, but, create object on a separate line and encode it here. Makes it easy for readability and debugging, just for future reference.

**Comment by shishirx34** [comment]
LGTM! :shipit:

**Comment by silhouettes** [comment]
This is great work, although while we're working on this area, I've been meaning to bring up several simplifications I'd really love to make to access keys in general, I'm hoping we can discuss this in greater detail

**Comment by geof90** [comment]
@lostintangent @dtivel @silhouettes I have edited the CLI to now take only a single parameter for logging in with access keys. Please take a look. :)

**Comment by lostintangent** [comment]
LGTM!

File: cli/README.md [hunk 3]
```diff
@@ -38,6 +38,19 @@ code-push logout
...
+code-push access-key add --des \"VSO Integration\"
+```
+
+You can generate additional access keys for your account with an optional description. This lets you log in to CodePush and manage your apps without going through the 3rd party authentication step, e.g. from a CI environment.
+
+```
+code-push login --key myKey
+```
+
+After generating the access key, you can call the login command with the generated key.
```
Comment (dtivel) [inline cli/README.md]: Move this sentence to the end of the previous paragraph.

File: cli/script/command-executor.ts [hunk 4]
```diff
@@ -21,13 +21,18 @@ import Promise = Q.Promise;
 
 var configFilePath: string = path.join(process.env.LOCALAPPDATA || process.env.HOME, ".code-push.config");
 
-interface IConnectionInfo {
+interface StandardLoginConnectionInfo {
```
Comment (dtivel) [inline cli/script/command-executor.ts]: IStandardLoginConnectionInfo

File: cli/script/command-executor.ts [hunk 5]
```diff
@@ -21,13 +21,18 @@ import Promise = Q.Promise;
...
     accessKeyName: string;
     providerName: string;
     providerUniqueId: string;
     serverUrl: string;
 }
 
+interface AccessKeyLoginConnectionInfo {
```
Comment (dtivel) [inline cli/script/command-executor.ts]: IAccessKeyLoginConnectionInfo

File: cli/script/command-executor.ts [hunk 6]
```diff
@@ -708,16 +759,37 @@ function requestAccessToken(): Promise<string> {
...
+        json = JSON.stringify(accessKeyLoginConnectionInfo);
+        fs.writeFileSync(configFilePath, json, { encoding: "utf8" });
+    }
 
-    json = JSON.stringify(info);
+    log("Login token persisted to file '" + configFilePath + "'. Run 'code-push logout' to remove the file.");
+}
 
-    fs.writeFileSync(configFilePath, json, { encoding: "utf8" });
+function tryBase64Decode(encoded: string) {
```
Comment (dtivel) [inline cli/script/command-executor.ts]: : string

File: cli/script/command-parser.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@ import * as semver from "semver";
 import * as chalk from "chalk";
 
 const USAGE_PREFIX = "Usage: code-push";
-const CODE_PUSH_URL = "https://codepush.azurewebsites.net";
+const CODE_PUSH_URL = "http://codepush.azurewebsites.net";
```
Comment (dtivel) [inline cli/script/command-parser.ts]: Why?

File: sdk/script/management/account-manager.ts [hunk 8]
```diff
@@ -60,40 +60,42 @@ export class AccountManager {
...
 
-            if (!loginInfo || !loginInfo.providerName || !loginInfo.providerUniqueId) {
-                reject(<CodePushError>{ message: "Invalid access key." });
-                return;
-            }
-
             var req = request.post(this.serverUrl + "/auth/login/accessToken");
-
             this.attachCredentials(req, request);
+            req = req.type("form")
```
Comment (dtivel) [inline sdk/script/management/account-manager.ts]: ;

File: cli/script/command-parser.ts [hunk 9]
```diff
@@ -29,6 +29,16 @@ function showHelp(showRootDescription?: boolean): void {
     }
 }
 
+function accessKeyAdd(commandName: string, yargs: yargs.Argv): void {
+    isValidCommand = true;
+    yargs.usage(USAGE_PREFIX + " access-key " + commandName + " --description <description>")
+        .demand(/*count*/ 2, /*max*/ 2)  // Require exactly two non-option arguments.
+        .example("access-key " + commandName + " --des \"VSO Integration\"", "Generates a new access key with the description \"VSO Integration\"")
+        .option("description", { alias: "des", demand: true, description: "A short description to be tagged to this access key", type: "string" });
```
Comment (dtivel) [inline cli/script/command-parser.ts]: Nowhere else do we use required options.  To date required options have been positional arguments.

File: cli/script/command-parser.ts (hunk 7 above)
Comment (geof90) [inline cli/script/command-parser.ts]: My mistake. I must have changed it when testing against localhost, and forgot to check that I changed it back to the right value.

File: cli/README.md [hunk 10]
```diff
@@ -36,13 +36,25 @@ code-push logout
 ```
 
 If you forget to logout from a machine you'd prefer not to leave a running session on (e.g. your friend's laptop), you can use the following commands to list and remove any "live" access tokens. 
+The list of access keys will display the name of the machine the token was created on, as well as the time the login occured. This should make it easy to spot keys you don't want to keep around.
```
Comment (dtivel) [inline cli/README.md]: occurred

File: cli/script/command-parser.ts [hunk 11]
```diff
@@ -29,6 +29,15 @@ export function showHelp(showRootDescription?: boolean): void {
     }
 }
 
+function accessKeyAdd(commandName: string, yargs: yargs.Argv): void {
+    isValidCommand = true;
+    yargs.usage(USAGE_PREFIX + " access-key " + commandName + " <description>")
+        .demand(/*count*/ 3, /*max*/ 3)  // Require exactly two non-option arguments.
+        .example("access-key " + commandName + " \"VSO Integration\"", "Generates a new access key with the description \"VSO Integration\"");
```
Comment (dtivel) [inline cli/script/command-parser.ts]: Why did you remove the "description" argument?

File: cli/README.md [hunk 12]
```diff
@@ -36,13 +36,25 @@ code-push logout
...
 If you forget to logout from a machine you'd prefer not to leave a running session on (e.g. your friend's laptop), you can use the following commands to list and remove any "live" access tokens. 
+The list of access keys will display the name of the machine the token was created on, as well as the time the login occurred. This should make it easy to spot keys you don't want to keep around.
 
 ```
 code-push access-key ls
 code-push access-key rm <accessKey>
 ```
 
-The list of access keys will display the name of the machine the token was created on, as well as the time the login occurred. This should make it easy to spot keys you don't want to keep around.
+You can generate additional access keys for your account with an optional description. This lets you log in to CodePush and manage your apps without going through the 3rd party authentication step, e.g. from a CI environment.
```
Comment (dtivel) [inline cli/README.md]: The description is required, I thought.

File: cli/README.md [hunk 13]
```diff
@@ -36,13 +36,25 @@ code-push logout
...
+
+```
+code-push access-key add
+code-push access-key add --des \"VSO Integration\"
```
Comment (dtivel) [inline cli/README.md]: Is this syntax correct?

File: cli/script/command-parser.ts [hunk 14]
```diff
@@ -190,9 +200,11 @@ var argv = yargs.usage(USAGE_PREFIX + " <command>")
     .command("login", "Authenticate with the CodePush server in order to begin managing your apps", (yargs: yargs.Argv) => {
         isValidCommandCategory = true;
         isValidCommand = true;
-        yargs.usage(USAGE_PREFIX + " login [serverUrl]")
+        yargs.usage(USAGE_PREFIX + " login [serverUrl] [--accessKey <accessKey>]")
             .demand(/*count*/ 1, /*max*/ 2)  // Require one non-optional and one optional argument.
             .example("login", "Logs in to " + CODE_PUSH_URL)
+            .example("login --key mykey", "Logs in on behalf of the user who owns and created the access key \"mykey\"")
```
Comment (dtivel) [inline cli/script/command-parser.ts]: --accessKey

File: cli/README.md [hunk 15]
```diff
@@ -36,13 +36,24 @@ code-push logout
...
+You can generate additional access keys for your account with a description tagged to each key. This lets you log in to CodePush and manage your apps without going through the 3rd party authentication step, e.g. from a CI environment.
+
+```
+code-push access-key add "VSO Integration"
+```
+
+After generating the access key, you can call the login command with the generated key.
+
+```
+code-push login --key myKey
```
Comment (dtivel) [inline cli/README.md]: --accessKey

File: cli/script/command-executor.ts [hunk 16]
```diff
@@ -551,19 +597,24 @@ function printList<T extends { id: string; name: string; }>(format: string, item
...
         log(JSON.stringify(dataSource));
     } else if (format === "table") {
-        printTable(["Key", "Description"], (dataSource: any[]): void => {
-            items.forEach((item: T): void => {
-                dataSource.push([item.name, item.description]);
+        printTable(["Key", "Time Created", "Created From", "Description"], (dataSource: any[]): void => {
+            keys.forEach((key: AccessKey): void => {
+                dataSource.push([
+                    key.name, 
+                    key.createdTime ? new Date(+key.createdTime).toString() : "",
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Exactly why is "+" required?

Comment (geof90) [inline cli/script/command-executor.ts]: The server was returning a string for the date number representation. This is fixed and no longer needed with the latest server change. :)

**Comment by dtivel** [comment]
LGTM
